### PR TITLE
feat: add buildStart buildEnd hook

### PR DESCRIPTION
**Pull request by underfin**
### Test Plan

---

### CODE REVIEW

**Comment by underfin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #208** <a href="https://app.graphite.dev/github/pr/rolldown-rs/rolldown/208" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/rolldown-rs/rolldown/208?utm_source=stack-comment).

File: packages/rolldown-core/src/rolldown-build.ts [hunk 1]
```diff
@@ -20,12 +21,17 @@ export class RolldownBuild implements RollupBuild {
   ): Promise<RolldownBuild> {
     // Convert `InputOptions` to `NormalizedInputOptions`.
     const normalizedInputOptions = await normalizeInputOptions(inputOptions)
+    await hookParallel('buildStart', [normalizedInputOptions], normalizedInputOptions.plugins)
```
Comment (hyf0) [inline packages/rolldown-core/src/rolldown-build.ts]: I intend to trigger this hook in rust side. Is this hook a blocking for pre-bundling? If not, let's not hurry to add it.

Comment (underfin) [reply packages/rolldown-core/src/rolldown-build.ts]: The `buildStart` hook moved to rust side, but the argument `options` as unmutable at js side. @hyf0 see see

File: crates/rolldown/src/bundler/graph/graph.rs [hunk 2]
```diff
@@ -24,20 +25,44 @@ pub struct Graph {
 }
 
 impl Graph {
-  pub async fn generate_module_graph<T: FileSystemExt + Default + 'static>(
+  pub async fn build<T: FileSystemExt + Default + 'static>(
```
Comment (underfin) [inline crates/rolldown/src/bundler/graph/graph.rs:28]: Add the `Graph.build` to easy compare to rollup code.

**Review by hyf0 — APPROVED** [review]

**Pull request merged**